File: frontend/src/features/thickness/logic/panels.ts
// Панели кузова и то, как статус панели превращается в цвет.
//
// Порогов здесь нет намеренно: статус (`factory` / `repaint` / `filler`) считает сервер
// и присылает готовым. Копия порога на клиенте — это вторая копия в вебе и третья в
// мобилке, и однажды они разойдутся на одном и том же замере.
import type { BodyPanel, PanelStatus } from '../../../shared/api/backend/thicknessContract'

export type PanelCode = BodyPanel

/** Статус панели плюс `none` — незамеренная панель. Провод её просто не присылает,
 *  а экран обязан показать: её отсутствие и есть то, что покупатель ищет. */
export type Grade = PanelStatus | 'none'

// Порядок списка — порядок осмотра машины, а не алфавит: сверху вниз и спереди назад.
export const PANELS: { code: PanelCode; label: string }[] = [
  { code: 'hood', label: 'Капот' },
  { code: 'roof', label: 'Крыша' },
  { code: 'trunk_lid', label: 'Крышка багажника' },
  { code: 'front_left_fender', label: 'Крыло переднее левое' },
  { code: 'front_right_fender', label: 'Крыло переднее правое' },
  { code: 'rear_left_fender', label: 'Крыло заднее левое' },
  { code: 'rear_right_fender', label: 'Крыло заднее правое' },
  { code: 'front_left_door', label: 'Дверь передняя левая' },
  { code: 'front_right_door', label: 'Дверь передняя правая' },
  { code: 'rear_left_door', label: 'Дверь задняя левая' },
  { code: 'rear_right_door', label: 'Дверь задняя правая' },
  { code: 'front_bumper', label: 'Бампер передний' },
  { code: 'rear_bumper', label: 'Бампер задний' },
]

export const PANEL_LABEL: Record<PanelCode, string> = Object.fromEntries(
  PANELS.map((panel) => [panel.code, panel.label]),
) as Record<PanelCode, string>

export const GRADE_COLOR: Record<Grade, string> = {
  factory: 'var(--measure-ok)',
  repaint: 'var(--measure-warn)',
  filler: 'var(--measure-bad)',
  none: 'var(--measure-none)',
}

export const GRADE_WORD: Record<Grade, string> = {
  factory: 'заводская',
  repaint: 'перекрашено',
  filler: 'шпаклёвка',
  none: 'не замерено',
}

// Подписи легенды. Числа — те, по которым считает сервер (спека
// `sale_car_thickness.yaml`); они здесь текст для человека, а не правило: цвет панели
// приходит статусом, и ни одна ветка кода не сравнивает число с этими границами.
export const GRADE_RANGE: Record<Grade, string | null> = {
  factory: 'меньше 200 мкм',
  repaint: '200–499',
  filler: 'от 500',
  none: null,
}

File: backend/app/features/listing/panels.py
"""Панели кузова и то, как читается замер толщиномера.

Набор панелей живёт здесь, а не в таблице: он не меняется от машины к машине, и
справочник из тринадцати строк, который никто не редактирует, — это таблица, которую
нужно наполнять миграцией, засевать в тестах и джойнить в каждом чтении карты.

Пороги тоже здесь и считаются на сервере. Клиент красит панель по `status`, а не по
числу: порог, посчитанный в вебе и в мобилке, — это две копии одного правила, и они
разойдутся на первой же правке.
"""

from enum import Enum


class BodyPanel(str, Enum):
    HOOD = "hood"
    ROOF = "roof"
    TRUNK_LID = "trunk_lid"
    FRONT_LEFT_DOOR = "front_left_door"
    FRONT_RIGHT_DOOR = "front_right_door"
    REAR_LEFT_DOOR = "rear_left_door"
    REAR_RIGHT_DOOR = "rear_right_door"
    FRONT_LEFT_FENDER = "front_left_fender"
    FRONT_RIGHT_FENDER = "front_right_fender"
    REAR_LEFT_FENDER = "rear_left_fender"
    REAR_RIGHT_FENDER = "rear_right_fender"
    FRONT_BUMPER = "front_bumper"
    REAR_BUMPER = "rear_bumper"


class PanelStatus(str, Enum):
    FACTORY = "factory"
    REPAINT = "repaint"
    FILLER = "filler"


TOTAL_PANELS = len(BodyPanel)

# Заводское покрытие держится в сотне с небольшим микрон; всё, что выше двухсот, —
# слой поверх заводского, а полтысячи и выше набирает только шпаклёвка.
REPAINT_FROM_UM = 200
FILLER_FROM_UM = 500

# Ноль и отрицательное прибор не показывает, а три тысячи выше любой шпаклёвки: за этими
# границами число пришло не от прибора.
MIN_VALUE_UM = 1
MAX_VALUE_UM = 3000


def status_of(value_um: int) -> PanelStatus:
    if value_um >= FILLER_FROM_UM:
        return PanelStatus.FILLER
    if value_um >= REPAINT_FROM_UM:
        return PanelStatus.REPAINT
    return PanelStatus.FACTORY
